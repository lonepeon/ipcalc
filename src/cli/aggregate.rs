use crate::cli::ErrorKind;
use crate::net::{CIDRParsingError, Mask, MaskParsingError, CIDR};

pub struct CLI<W: std::io::Write> {
    pub out: W,
}

impl<W: std::io::Write> CLI<W> {
    pub fn new(out: W) -> Self {
        CLI { out }
    }

    pub fn execute(&mut self, raw_cidr: String, raw_mask: String) -> Result<(), ErrorKind> {
        if raw_cidr.is_empty() {
            return Err(ErrorKind::InvalidInput(
                "expecting non empty CIDR argument".to_string(),
            ));
        }

        if raw_mask.is_empty() {
            return Err(ErrorKind::InvalidInput(
                "expecting non empty MASK argument".to_string(),
            ));
        }

        let mask = match raw_mask.parse::<Mask>() {
            Ok(mask) => mask,
            Err(MaskParsingError::InvalidRange) => {
                return Err(ErrorKind::InvalidInput(
                    "invalid mask range value".to_string(),
                ))
            }
            Err(MaskParsingError::InvalidFormat) => {
                return Err(ErrorKind::InvalidInput("invalid mask format".to_string()))
            }
        };

        let cidr = match raw_cidr.parse::<CIDR>() {
            Ok(cidr) => cidr,
            Err(CIDRParsingError::InvalidMaskLength) => {
                return Err(ErrorKind::InvalidInput(
                    "IPv4 CIDR masklength must be between 0 and 32".to_string(),
                ));
            }
            Err(CIDRParsingError::InvalidHostFormat) => {
                return Err(ErrorKind::InvalidInput(
                    "invalid IPv4 CIDR format".to_string(),
                ))
            }
        };

        let start = mask.prefix_length().min(cidr.mask().prefix_length());
        let stop = mask.prefix_length().max(cidr.mask().prefix_length());
        for mask_length in start..=stop {
            let mask = Mask::new(mask_length).expect("{} is not a valid mask length");
            writeln!(self.out, "{}", cidr.aggregate(mask)).unwrap();
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn aggregate_32_to_24() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        cli.execute("10.12.5.255/32".to_string(), "24".to_string())
            .unwrap();

        let expected_output = include_str!("testdata/aggregate-32-to-24.golden");
        let actual_output = String::from_utf8(output).unwrap();

        assert_eq!(expected_output, actual_output);
    }

    #[test]
    fn aggregate_24_to_32() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        cli.execute("10.12.5.255/24".to_string(), "32".to_string())
            .unwrap();

        let expected_output = include_str!("testdata/aggregate-24-to-32.golden");
        let actual_output = String::from_utf8(output).unwrap();

        assert_eq!(expected_output, actual_output);
    }
}
