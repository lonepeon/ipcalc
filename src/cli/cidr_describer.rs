use crate::net::CIDR;
use core::fmt;

pub struct CIDRDescriber {
    pub cidr: CIDR,
    pub with_binary: bool,
}

impl fmt::Display for CIDRDescriber {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Address:   {:15}      ", format!("{}", self.cidr.ip()))?;
        if self.with_binary {
            write!(f, "{:b}", self.cidr.ip())?;
        }
        writeln!(f)?;

        write!(
            f,
            "Netmask:   {:20} ",
            format!("{} = {}", self.cidr.mask(), self.cidr.mask().len())
        )?;
        if self.with_binary {
            write!(f, "{:b}", self.cidr.mask())?;
        }
        writeln!(f)?;

        write!(
            f,
            "Wildcard:  {:15}      ",
            format!("{}", self.cidr.wildcard_mask())
        )?;
        if self.with_binary {
            write!(f, "{:b}", self.cidr.wildcard_mask())?;
        }
        writeln!(f)?;

        writeln!(f, "=>")?;

        write!(f, "Network:   {:18}   ", format!("{}", self.cidr.network()))?;
        if self.with_binary {
            write!(f, "{:b}", self.cidr.network().ip())?;
        }
        writeln!(f)?;

        write!(
            f,
            "HostMin:   {:18}   ",
            self.cidr
                .first_address()
                .map(|ip| format!("{}", ip))
                .unwrap_or_else(|| "n/a".to_string()),
        )?;
        if self.with_binary {
            self.cidr.first_address().map(|ip| write!(f, "{:b}", ip));
        }
        writeln!(f)?;

        write!(
            f,
            "HostMax:   {:18}   ",
            self.cidr
                .last_address()
                .map(|ip| format!("{}", ip))
                .unwrap_or_else(|| "n/a".to_string()),
        )?;
        if self.with_binary {
            self.cidr.last_address().map(|ip| write!(f, "{:b}", ip));
        }
        writeln!(f)?;

        write!(
            f,
            "Broadcast: {:18}   ",
            self.cidr
                .broadcast_address()
                .map(|ip| format!("{}", ip))
                .unwrap_or_else(|| "n/a".to_string()),
        )?;
        if self.with_binary {
            self.cidr
                .broadcast_address()
                .map(|ip| write!(f, "{:b}", ip));
        }
        writeln!(f)?;

        write!(
            f,
            "Hosts/Net: {:10}           class {}, {}",
            format!("{}", self.cidr.hosts()),
            self.cidr.class(),
            self.cidr.kind(),
        )?;

        writeln!(f)
    }
}
