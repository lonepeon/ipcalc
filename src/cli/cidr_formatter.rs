use crate::net::CIDR;
use core::fmt;

pub struct CIDRFormatter {
    pub cidr: CIDR,
    pub with_binary: bool,
}

impl fmt::Display for CIDRFormatter {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Address:   {:15}      ", format!("{}", self.cidr.ip()))?;
        if self.with_binary {
            write!(f, "{:b}", self.cidr.ip())?;
        }
        writeln!(f)?;

        write!(
            f,
            "Netmask:   {:20} ",
            format!(
                "{} = {}",
                self.cidr.mask(),
                self.cidr.mask().prefix_length()
            )
        )?;
        if self.with_binary {
            write!(f, "{:b}", self.cidr.mask())?;
        }
        writeln!(f)?;

        write!(
            f,
            "Wildcard:  {:15}      ",
            format!("{}", self.cidr.wildcard_mask())
        )?;
        if self.with_binary {
            write!(f, "{:b}", self.cidr.wildcard_mask())?;
        }
        writeln!(f)?;

        writeln!(f, "=>")?;

        write!(
            f,
            "Network:   {:18}   ",
            format!("{}", self.cidr.network_address())
        )?;
        if self.with_binary {
            write!(f, "{:b}", self.cidr.network_address().ip())?;
        }
        writeln!(f)?;

        write!(
            f,
            "HostMin:   {:18}   ",
            self.cidr
                .first_address()
                .map(|ip| format!("{}", ip))
                .unwrap_or_else(|| "n/a".to_string()),
        )?;
        if self.with_binary {
            self.cidr.first_address().map(|ip| write!(f, "{:b}", ip));
        }
        writeln!(f)?;

        write!(
            f,
            "HostMax:   {:18}   ",
            self.cidr
                .last_address()
                .map(|ip| format!("{}", ip))
                .unwrap_or_else(|| "n/a".to_string()),
        )?;
        if self.with_binary {
            self.cidr.last_address().map(|ip| write!(f, "{:b}", ip));
        }
        writeln!(f)?;

        write!(
            f,
            "Broadcast: {:18}   ",
            self.cidr
                .broadcast_address()
                .map(|ip| format!("{}", ip))
                .unwrap_or_else(|| "n/a".to_string()),
        )?;
        if self.with_binary {
            self.cidr
                .broadcast_address()
                .map(|ip| write!(f, "{:b}", ip));
        }
        writeln!(f)?;

        write!(
            f,
            "Hosts/Net: {:10}           class {}, {}",
            format!("{}", self.cidr.hosts()),
            self.cidr.class(),
            self.cidr.kind(),
        )?;

        writeln!(f)
    }
}
