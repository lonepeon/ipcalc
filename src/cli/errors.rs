#[derive(Debug)]
pub enum ErrorKind {
    InvalidInput(String),
}
