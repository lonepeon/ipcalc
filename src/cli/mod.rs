pub mod describe;
mod errors;

pub use self::errors::ErrorKind;
