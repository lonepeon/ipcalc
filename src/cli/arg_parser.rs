use crate::cli::ErrorKind;
use crate::net::{CIDRParsingError, Mask, MaskParsingError, CIDR};

pub fn parse_cidr(name: &str, raw: String) -> Result<CIDR, ErrorKind> {
    if raw.is_empty() {
        return Err(ErrorKind::InvalidInput(format!(
            "expecting non empty {} argument",
            name
        )));
    }

    match raw.parse::<CIDR>() {
        Ok(cidr) => Ok(cidr),
        Err(CIDRParsingError::InvalidMaskLength) => Err(ErrorKind::InvalidInput(
            "masklength must be between 0 and 32".to_string(),
        )),
        Err(CIDRParsingError::InvalidHostFormat) => Err(ErrorKind::InvalidInput(
            "invalid IPv4 CIDR format".to_string(),
        )),
    }
}

pub fn parse_mask(name: &str, raw: String) -> Result<Mask, ErrorKind> {
    if raw.is_empty() {
        return Err(ErrorKind::InvalidInput(format!(
            "expecting non empty {} argument",
            name
        )));
    }

    match raw.parse::<Mask>() {
        Ok(mask) => Ok(mask),
        Err(MaskParsingError::InvalidRange) => Err(ErrorKind::InvalidInput(
            "mask length must be between 0 and 32".to_string(),
        )),
        Err(MaskParsingError::InvalidFormat) => {
            Err(ErrorKind::InvalidInput("invalid mask format".to_string()))
        }
    }
}
