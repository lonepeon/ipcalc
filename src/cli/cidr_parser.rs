use crate::cli::ErrorKind;
use crate::net::{CIDRParsingError, CIDR};

pub fn parser(raw: String) -> Result<CIDR, ErrorKind> {
    match raw.parse::<CIDR>() {
        Ok(cidr) => Ok(cidr),
        Err(CIDRParsingError::InvalidMaskLength) => Err(ErrorKind::InvalidInput(
            "masklength must be between 0 and 32".to_string(),
        )),
        Err(CIDRParsingError::InvalidHostFormat) => Err(ErrorKind::InvalidInput(
            "invalid IPv4 CIDR format".to_string(),
        )),
    }
}
