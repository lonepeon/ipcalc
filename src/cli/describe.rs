use crate::net::{CIDRParsingError, CIDR};

use crate::cli::ErrorKind;

pub struct CLI<W: std::io::Write> {
    pub out: W,
    pub with_binary: bool,
}

impl<W: std::io::Write> CLI<W> {
    pub fn new(out: W) -> Self {
        CLI {
            out,
            with_binary: true,
        }
    }

    fn write<D: std::fmt::Display>(&mut self, d: D) {
        write!(&mut self.out, "{}", d).unwrap();
    }

    fn write_if_binary<D: std::fmt::Display>(&mut self, d: D) {
        if !self.with_binary {
            return;
        }
        write!(&mut self.out, "{}", d).unwrap();
    }

    pub fn execute(&mut self, args: Vec<String>) -> Result<(), ErrorKind> {
        if args.len() != 1 {
            return Err(ErrorKind::InvalidInput(
                "expecting only one CIDR as argument".to_string(),
            ));
        }

        match args[0].parse::<CIDR>() {
            Ok(cidr) => {
                self.write(format!("Address:   {:15}      ", format!("{}", cidr.ip())));
                self.write_if_binary(format!("{:b}", cidr.ip()));
                self.write("\n");

                self.write(format!(
                    "Netmask:   {:20} ",
                    format!("{} = {}", cidr.mask(), cidr.mask().len())
                ));
                self.write_if_binary(format!("{:b}", cidr.mask()));
                self.write("\n");

                self.write(format!(
                    "Wildcard:  {:15}      ",
                    format!("{}", cidr.wildcard_mask())
                ));
                self.write_if_binary(format!("{:b}", cidr.wildcard_mask()));
                self.write("\n");

                self.write("=>\n");

                self.write(format!(
                    "Network:   {:18}   ",
                    format!("{}", cidr.network())
                ));
                self.write_if_binary(format!("{:b}", cidr.network().ip()));
                self.write("\n");

                self.write(format!(
                    "HostMin:   {:18}   ",
                    cidr.first_address()
                        .map(|ip| format!("{}", ip))
                        .unwrap_or_else(|| "n/a".to_string()),
                ));
                self.write_if_binary(
                    cidr.first_address()
                        .map(|ip| format!("{:b}", ip))
                        .unwrap_or_else(|| "".to_string()),
                );
                self.write("\n");

                self.write(format!(
                    "HostMax:   {:18}   ",
                    cidr.last_address()
                        .map(|ip| format!("{}", ip))
                        .unwrap_or_else(|| "n/a".to_string()),
                ));
                self.write_if_binary(
                    cidr.last_address()
                        .map(|ip| format!("{:b}", ip))
                        .unwrap_or_else(|| "".to_string()),
                );
                self.write("\n");

                self.write(format!(
                    "Broadcast: {:18}   ",
                    cidr.broadcast_address()
                        .map(|ip| format!("{}", ip))
                        .unwrap_or_else(|| "n/a".to_string()),
                ));
                self.write_if_binary(
                    cidr.broadcast_address()
                        .map(|ip| format!("{:b}", ip))
                        .unwrap_or_else(|| "".to_string()),
                );
                self.write("\n");

                self.write(format!(
                    "Hosts/Net: {:10}           class {}, {}",
                    format!("{}", cidr.hosts()),
                    cidr.class(),
                    cidr.kind(),
                ));
                self.write("\n");

                Ok(())
            }
            Err(CIDRParsingError::InvalidMaskLength) => Err(ErrorKind::InvalidInput(
                "masklength must be between 0 and 32".to_string(),
            )),
            Err(CIDRParsingError::InvalidHostFormat) => Err(ErrorKind::InvalidInput(
                "host must use aaa.bbb.ccc.ddd format".to_string(),
            )),
        }
    }
}

#[cfg(test)]
mod tests {
    use std::fs;

    #[test]
    fn describe_class_a() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        cli.execute(vec!["10.12.23.43/20".to_string()]).unwrap();

        let expected_output = fs::read_to_string("src/cli/testdata/describe.golden").unwrap();
        let actual_output = String::from_utf8(output).unwrap();

        assert_eq!(expected_output, actual_output);
    }
}
