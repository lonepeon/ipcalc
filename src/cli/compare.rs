use crate::cli::arg_parser;
use crate::cli::ErrorKind;
use crate::net::CIDRComparison;

pub struct CLI<W: std::io::Write> {
    pub out: W,
}

impl<W: std::io::Write> CLI<W> {
    pub fn new(out: W) -> Self {
        CLI { out }
    }

    pub fn execute(&mut self, raw_cidr: String, raw_other: String) -> Result<(), ErrorKind> {
        let cidr = arg_parser::parse_cidr("CIDR", raw_cidr)?;
        let other = arg_parser::parse_cidr("OTHER_CIDR", raw_other)?;
        let description = match cidr.compare(&other) {
            CIDRComparison::Subset => "is a subset of",
            CIDRComparison::Superset => "is a superset of",
            CIDRComparison::Equals => "is the same as",
            CIDRComparison::Different => "is in a different network than",
        };

        writeln!(self.out, "{} {} {}", cidr, description, other,).unwrap();

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use crate::cli::ErrorKind;

    #[test]
    fn compare_empty_cidr1() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        let output = cli.execute("".to_string(), "10.12.25.99/24".to_string());

        assert_eq!(
            Err(ErrorKind::InvalidInput(
                "expecting non empty CIDR argument".to_string()
            )),
            output
        );
    }

    #[test]
    fn compare_unparsable_cidr1() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        let output = cli.execute("not a CIDR".to_string(), "10.12.25.99/24".to_string());

        assert_eq!(
            Err(ErrorKind::InvalidInput(
                "invalid IPv4 CIDR format".to_string()
            )),
            output
        );
    }

    #[test]
    fn compare_invalid_mask_cidr1() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        let output = cli.execute("10.12.23.43/200".to_string(), "10.12.25.99/24".to_string());

        assert_eq!(
            Err(ErrorKind::InvalidInput(
                "masklength must be between 0 and 32".to_string()
            )),
            output
        );
    }

    #[test]
    fn compare_empty_cidr2() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        let output = cli.execute("10.12.23.43/16".to_string(), "".to_string());

        assert_eq!(
            Err(ErrorKind::InvalidInput(
                "expecting non empty OTHER_CIDR argument".to_string()
            )),
            output
        );
    }

    #[test]
    fn compare_unparsable_cidr2() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        let output = cli.execute("10.12.23.43/16".to_string(), "not a CIDR".to_string());

        assert_eq!(
            Err(ErrorKind::InvalidInput(
                "invalid IPv4 CIDR format".to_string()
            )),
            output
        );
    }
    #[test]
    fn compare_invalid_mask_cidr2() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        let output = cli.execute("10.12.23.43/16".to_string(), "10.12.25.99/200".to_string());

        assert_eq!(
            Err(ErrorKind::InvalidInput(
                "masklength must be between 0 and 32".to_string()
            )),
            output
        );
    }

    #[test]
    fn on_superset() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        cli.execute("10.12.23.43/16".to_string(), "10.12.25.99/24".to_string())
            .unwrap();

        let expected_output = include_str!("testdata/compare-superset.golden");
        let actual_output = String::from_utf8(output).unwrap();

        assert_eq!(expected_output, actual_output);
    }

    #[test]
    fn on_subset() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        cli.execute("10.12.23.43/24".to_string(), "10.12.25.99/16".to_string())
            .unwrap();

        let expected_output = include_str!("testdata/compare-subset.golden");
        let actual_output = String::from_utf8(output).unwrap();

        assert_eq!(expected_output, actual_output);
    }

    #[test]
    fn on_equals() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        cli.execute("10.12.23.43/16".to_string(), "10.12.25.99/16".to_string())
            .unwrap();

        let expected_output = include_str!("testdata/compare-equals.golden");
        let actual_output = String::from_utf8(output).unwrap();

        assert_eq!(expected_output, actual_output);
    }

    #[test]
    fn on_different() {
        let mut output = Vec::new();
        let mut cli = super::CLI::new(&mut output);
        cli.execute("10.12.23.43/16".to_string(), "10.14.25.99/16".to_string())
            .unwrap();

        let expected_output = include_str!("testdata/compare-different.golden");
        let actual_output = String::from_utf8(output).unwrap();

        assert_eq!(expected_output, actual_output);
    }
}
