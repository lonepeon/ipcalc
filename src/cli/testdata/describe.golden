Address:   10.12.23.43          00001010.00001100.00010111.00101011
Netmask:   255.255.240.0 = 20   11111111.11111111.11110000.00000000
Wildcard:  0.0.15.255           00000000.00000000.00001111.11111111
=>
Network:   10.12.16.0/20        00001010.00001100.00010000.00000000
HostMin:   10.12.16.1           00001010.00001100.00010000.00000001
HostMax:   10.12.31.254         00001010.00001100.00011111.11111110
Broadcast: 10.12.31.255         00001010.00001100.00011111.11111111
Hosts/Net: 4094                 class A, Private Internet
